kotlinx-coroutines-core/jvm/src/CancellationPoint.kt

```diff
@@ -0,0 +1,155 @@
+ * suspend fun <T> BlockingQueue<T>.awaitTake(): T =
+ *         runInterruptible(Dispatchers.IO) { queue.take() }
+ * ```
+ *
+ * @param context additional to [CoroutineScope.coroutineContext] context of the coroutine.
+ * @param block regular blocking block that will be interrupted on coroutine cancellation.
+ */
+public suspend fun <T> runInterruptible(
+        context: CoroutineContext = EmptyCoroutineContext,
+        block: InterruptibleScope.() -> T
+): T {
+    // fast path: empty context
+    if (context === EmptyCoroutineContext) { return runInterruptibleInExpectedContext(block) }
+    // slow path:
+    return withContext(context) { runInterruptibleInExpectedContext(block) }
+}
+
+/**
+ * Defines a scope for new interruptible blocking code blocks.
+ *
+ * Currently, The main goal is to restrict access to outer code scope methods (using outer scopes as
+ * implicit receivers). See [CodeScope] for details.
+ */
+public interface InterruptibleScope: CodeScope
+
+private suspend fun <T> runInterruptibleInExpectedContext(block: InterruptibleScope.() -> T): T =
+        suspendCoroutineUninterceptedOrReturn sc@{ uCont ->
+            try {
+                // fast path: no job
+                val job = uCont.context[Job] ?: return@sc DefaultInterruptibleScope.block()
+                // slow path
+                val threadState = ThreadState().apply { initInterrupt(job) }
+                try {
+                    DefaultInterruptibleScope.block()
+                } finally {
+                    threadState.clearInterrupt()
+                }
+            } catch (e: InterruptedException) {
+                throw CancellationException()
+            }
+        }
+
+private object DefaultInterruptibleScope : InterruptibleScope
+
+private class ThreadState {
+
+    fun initInterrupt(job: Job) {
+        // starts with Init
+        if (state.value !== Init) throw IllegalStateException("impossible state")
+        // remembers this running thread
+        state.value = Working(Thread.currentThread(), null)
+        // watches the job for cancellation
+        val cancelHandle =
+                job.invokeOnCompletion(onCancelling = true, invokeImmediately = true, handler = CancelHandler())
+        // remembers the cancel handle or drops it
+        state.loop { s ->
+            when {
+                s is Working -> if (state.compareAndSet(s, Working(s.thread, cancelHandle))) return
+                s === Interrupting || s === Interrupted -> return
+                s === Init || s === Finish -> throw IllegalStateException("impossible state")
+                else -> throw IllegalStateException("unknown state")
+            }
+        }
+    }
+
+    fun clearInterrupt() {
+        state.loop { s ->
+            when {
+                s is Working -> if (state.compareAndSet(s, Finish)) { s.cancelHandle!!.dispose(); return }
+                s === Interrupting -> Thread.yield() // eases the thread
```

Comment:
Please avoid `Thread.yield`. It is way too heavy-weight for 99% of usages.
Just spinning is enough